src/bun.js/child_process.exports.js

```diff
@@ -1,3 +1,995 @@
+//       } else {
+//         args = ["-c", command];
+//       }
+//     } else {
+//       if (typeof options.shell === "string") file = options.shell;
+//       else if (process.platform === "android") file = "/system/bin/sh";
+//       else file = "/bin/sh";
+//       args = ["-c", command];
+//     }
+//   }
+
+//   if (typeof options.argv0 === "string") {
+//     ArrayPrototypeUnshift(args, options.argv0);
+//   } else {
+//     ArrayPrototypeUnshift(args, file);
+//   }
+
+//   const env = options.env || process.env;
+//   const envPairs = [];
+
+//   // process.env.NODE_V8_COVERAGE always propagates, making it possible to
+//   // collect coverage for programs that spawn with white-listed environment.
+//   copyProcessEnvToEnv(env, "NODE_V8_COVERAGE", options.env);
+
+//   let envKeys = [];
+//   // Prototype values are intentionally included.
+//   for (const key in env) {
+//     ArrayPrototypePush(envKeys, key);
+//   }
+
+//   if (process.platform === "win32") {
+//     // On Windows env keys are case insensitive. Filter out duplicates,
+//     // keeping only the first one (in lexicographic order)
+//     const sawKey = new SafeSet();
+//     envKeys = ArrayPrototypeFilter(ArrayPrototypeSort(envKeys), (key) => {
+//       const uppercaseKey = StringPrototypeToUpperCase(key);
+//       if (sawKey.has(uppercaseKey)) {
+//         return false;
+//       }
+//       sawKey.add(uppercaseKey);
+//       return true;
+//     });
+//   }
+
+//   for (const key of envKeys) {
+//     const value = env[key];
+//     if (value !== undefined) {
+//       validateArgumentNullCheck(key, `options.env['${key}']`);
+//       validateArgumentNullCheck(value, `options.env['${key}']`);
+//       ArrayPrototypePush(envPairs, `${key}=${value}`);
+//     }
+//   }
+
+//   return {
+//     // Make a shallow copy so we don't clobber the user's options object.
+//     ...options,
+//     args,
+//     cwd,
+//     detached: !!options.detached,
+//     envPairs,
+//     file,
+//     // windowsHide: !!options.windowsHide,
+//     // windowsVerbatimArguments: !!windowsVerbatimArguments,
+//   };
+// }
+
+//------------------------------------------------------------------------------
+// Section 3. ChildProcess class
+//------------------------------------------------------------------------------
+
+export class ChildProcess extends EventEmitter {
+  constructor() {
+    super();
+    this._closesNeeded = 0;
+    this._closesGot = 0;
+    this.connected = false;
+    this.signalCode = null;
+    this.exitCode = null;
+    this.killed = false;
+    this.spawnfile = undefined;
+    this.spawnargs = undefined;
+    this.pid = undefined;
+    this.stdin = undefined;
+    this.stdout = undefined;
+    this.stderr = undefined;
+    this.stdio = undefined;
+    this.channel = undefined;
+    this._handle = undefined;
+    this._handleQueue = undefined;
+    this._pendingMessage = undefined;
+    this._pendingHandle = undefined;
+    this._channel = undefined;
```

Comment:
Are npm packages using these? Can you check using source graph or cs.github.com?

Any internal private variables should use #private fields so they're fully inaccessible

Also, it'd be best to declare all the fields in the class definition to hint to the engine what fields to expect.

Reply:
Yes I'll go ahead and check. I was mostly ripping these from the node implementation which I'm assuming was written before the addition of true private fields. Hopefully no one is using these though.

I'll make sure they get defined in the class definition 👍